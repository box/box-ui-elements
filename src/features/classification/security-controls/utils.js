// @flow
import getProp from 'lodash/get';
import isNil from 'lodash/isNil';

import type { Controls, MessageItem } from '../flowTypes';

import appRestrictionsMessageMap from './appRestrictionsMessageMap';
import downloadRestrictionsMessageMap from './downloadRestrictionsMessageMap';
import messages from './messages';
import {
    ACCESS_POLICY_RESTRICTION,
    APP_RESTRICTION_MESSAGE_TYPE,
    DOWNLOAD_CONTROL,
    LIST_ACCESS_LEVEL,
    SHARED_LINK_ACCESS_LEVEL,
} from '../constants';

const { SHARED_LINK, DOWNLOAD, EXTERNAL_COLLAB, APP, WATERMARK } = ACCESS_POLICY_RESTRICTION;
const { DEFAULT, WITH_APP_LIST, WITH_OVERFLOWN_APP_LIST } = APP_RESTRICTION_MESSAGE_TYPE;
const { DESKTOP, MOBILE, WEB } = DOWNLOAD_CONTROL;
const { BLOCK, WHITELIST, BLACKLIST } = LIST_ACCESS_LEVEL;
const { COLLAB_ONLY, COLLAB_AND_COMPANY_ONLY, PUBLIC } = SHARED_LINK_ACCESS_LEVEL;

const getShortSecurityControlsMessage = (controls: Controls): Array<MessageItem> => {
    const items = [];
    const { sharedLink, download, externalCollab, app, watermark } = controls;

    // Shared link and external collab restrictions are grouped
    // together as generic "sharing" restrictions
    const sharing = (sharedLink && sharedLink.accessLevel !== PUBLIC) || externalCollab;

    if (sharing && download && app) {
        items.push({ message: messages.shortAllRestrictions });
    } else if (sharing && download) {
        items.push({ message: messages.shortSharingDownload });
    } else if (sharing && app) {
        items.push({ message: messages.shortSharingApp });
    } else if (download && app) {
        items.push({ message: messages.shortDownloadApp });
    } else if (sharing) {
        items.push({ message: messages.shortSharing });
    } else if (download) {
        items.push({ message: messages.shortDownload });
    } else if (app) {
        items.push({ message: messages.shortApp });
    }

    if (watermark) {
        items.push({ message: messages.shortWatermarking });
    }

    return items;
};

const getSharedLinkMessages = (controls: Controls): Array<MessageItem> => {
    const items = [];
    const accessLevel = getProp(controls, `${SHARED_LINK}.accessLevel`);

    switch (accessLevel) {
        case COLLAB_ONLY:
            items.push({ message: messages.sharingCollabOnly });
            break;
        case COLLAB_AND_COMPANY_ONLY:
            items.push({ message: messages.sharingCollabAndCompanyOnly });
            break;
        default:
            // no-op
            break;
    }
    return items;
};

const getWatermarkingMessages = (controls: Controls): Array<MessageItem> => {
    const items = [];
    const isWatermarkEnabled = getProp(controls, `${WATERMARK}.enabled`, false);
    if (isWatermarkEnabled) {
        items.push({ message: messages.watermarkingApplied });
    }

    return items;
};

const getExternalCollabMessages = (controls: Controls): Array<MessageItem> => {
    const items = [];
    const accessLevel = getProp(controls, `${EXTERNAL_COLLAB}.accessLevel`);

    switch (accessLevel) {
        case BLOCK:
            items.push({ message: messages.externalCollabBlock });
            break;
        case WHITELIST:
        case BLACKLIST:
            items.push({ message: messages.externalCollabDomainList });
            break;
        default:
            // no-op
            break;
    }
    return items;
};

const getAppDownloadMessages = (controls: Controls, maxAppCount?: number): Array<MessageItem> => {
    const items = [];
    const accessLevel = getProp(controls, `${APP}.accessLevel`);

    switch (accessLevel) {
        case BLOCK:
            items.push({ message: messages.appDownloadRestricted });
            break;
        case WHITELIST:
        case BLACKLIST: {
            const apps = getProp(controls, `${APP}.apps`, []);

            maxAppCount = isNil(maxAppCount) ? apps.length : maxAppCount;
            const appsToDisplay = apps.slice(0, maxAppCount);
            const remainingAppCount = apps.slice(maxAppCount).length;
            const appNames = appsToDisplay.map(({ displayText }) => displayText).join(', ');

            if (remainingAppCount) {
                const appsList = apps.map(({ displayText }) => displayText).join(', ');

                items.push({
                    message: {
                        ...appRestrictionsMessageMap[accessLevel][WITH_OVERFLOWN_APP_LIST],
                        values: { appNames, remainingAppCount },
                    },
                    tooltipMessage: {
                        ...messages.allAppNames,
                        values: { appsList },
                    },
                });
            } else {
                // Display list of apps if available, otherwise use generic
                // app restriction copy
                const messageType = apps.length ? WITH_APP_LIST : DEFAULT;

                items.push({
                    message: {
                        ...appRestrictionsMessageMap[accessLevel][messageType],
                        values: { appNames },
                    },
                });
            }
            break;
        }
        default:
            // no-op
            break;
    }
    return items;
};

const getDownloadMessages = (controls: Controls): Array<MessageItem> => {
    const items = [];
    const { web, mobile, desktop } = getProp(controls, DOWNLOAD, {});

    const downloadRestrictions = {
        [WEB]: {
            platform: WEB,
            restrictions: web,
        },
        [MOBILE]: {
            platform: MOBILE,
            restrictions: mobile,
        },
        [DESKTOP]: {
            platform: DESKTOP,
            restrictions: desktop,
        },
    };

    Object.keys(downloadRestrictions).forEach(platformKey => {
        const { platform, restrictions } = downloadRestrictions[platformKey];

        if (!restrictions) {
            return;
        }
        const { restrictExternalUsers, restrictManagedUsers } = restrictions;

        if (restrictManagedUsers && restrictExternalUsers) {
            items.push({ message: downloadRestrictionsMessageMap[platform].externalRestricted[restrictManagedUsers] });
        } else if (restrictManagedUsers) {
            items.push({ message: downloadRestrictionsMessageMap[platform].externalAllowed[restrictManagedUsers] });
        } else if (restrictExternalUsers) {
            items.push({ message: downloadRestrictionsMessageMap[platform].externalRestricted.default });
        }
    });

    return items;
};

const getFullSecurityControlsMessages = (controls: Controls, maxAppCount?: number): Array<MessageItem> => {
    const items = [
        ...getSharedLinkMessages(controls),
        ...getExternalCollabMessages(controls),
        ...getDownloadMessages(controls),
        ...getAppDownloadMessages(controls, maxAppCount),
        ...getWatermarkingMessages(controls),
    ];
    return items;
};

export { getShortSecurityControlsMessage, getFullSecurityControlsMessages };
