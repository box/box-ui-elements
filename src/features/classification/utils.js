// @flow
import getProp from 'lodash/get';

import classificationColorsMap from './classificationColorsMap';
import { DEFAULT_CLASSIFICATION_COLOR_ID } from './constants';

function getClassificationLabelColors(classificationLabelData?: ?{ colorID?: number }) {
    const colorID = getProp(classificationLabelData, 'colorID', DEFAULT_CLASSIFICATION_COLOR_ID);
    const { fillColor, strokeColor } =
        classificationColorsMap[colorID] || classificationColorsMap[DEFAULT_CLASSIFICATION_COLOR_ID];

    return { fillColor, strokeColor };
}

// eslint-disable-next-line import/prefer-default-export
export { getClassificationLabelColors };
