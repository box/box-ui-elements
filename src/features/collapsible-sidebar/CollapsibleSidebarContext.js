// @flow
import * as React from 'react';

export default React.createContext<{ isScrolling: boolean }>({ isScrolling: false });
