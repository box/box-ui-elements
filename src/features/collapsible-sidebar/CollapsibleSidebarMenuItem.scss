@import '../../styles/variables';

.bdl-CollapsibleSidebar-menuItemLink {
    transition: background-color .15s cubic-bezier(.215, .61, .355, 1);

    a:hover &,
    a:active &,
    a:focus &,
    .is-currentPage & {
        .bdl-CollapsibleSidebar-menuItemIcon,
        .bdl-CollapsibleSidebar-menuItemLabel {
            opacity: 1;
        }
    }

    // Focus state handled by styled-component theme
    &:focus {
        outline: none;
    }
}

.bdl-CollapsibleSidebar-menuItem {
    display: flex;
    align-items: center;
    width: 100%;
    height: 10 * $bdl-grid-unit;
    padding: 0 3 * $bdl-grid-unit;
    overflow-x: hidden;
    font-weight: bold;
    white-space: nowrap;
    border: 1px solid transparent;
    border-radius: $bdl-grid-unit * 2;
}

.bdl-CollapsibleSidebar-menuItemLabel {
    flex-grow: 1;
    overflow: hidden;
    text-overflow: ellipsis;
    opacity: .85;
}

.bdl-CollapsibleSidebar-menuItemIcon {
    position: relative;
    top: 1px;
    opacity: .7;
}

.bdl-CollapsibleSidebar-menuItemIcon + .bdl-CollapsibleSidebar-menuItemLabel {
    margin-left: $bdl-grid-unit * 4;
}
