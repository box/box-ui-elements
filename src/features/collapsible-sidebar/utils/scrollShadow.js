// @flow
import classNames from 'classnames';

import './scrollShadow.scss';

type scrollShadowState = {
    isBottomOverflowed?: boolean,
    isTopOverflowed?: boolean,
};

const getScrollShadowState = (scrollTop: number, scrollHeight: number, clientHeight: number): scrollShadowState => {
    const scrollState: scrollShadowState = {};

    if (scrollTop > 0) {
        scrollState.isTopOverflowed = true;
    }

    if (scrollHeight - clientHeight > scrollTop) {
        scrollState.isBottomOverflowed = true;
    }

    return scrollState;
};

const getScrollShadowClassName = (scrollTop: number, scrollHeight: number, clientHeight: number): string => {
    const scrollState = getScrollShadowState(scrollTop, scrollHeight, clientHeight);

    return classNames('scroll-shadow-container', {
        'is-showing-top-shadow': scrollState.isTopOverflowed,
        'is-showing-bottom-shadow': scrollState.isBottomOverflowed,
    });
};

export { getScrollShadowClassName, getScrollShadowState };
