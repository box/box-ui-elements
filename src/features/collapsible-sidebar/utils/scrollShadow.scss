// Constraints for the shadow shape
$scroll-shadow-height: 6px;

.scroll-shadow-container {
    &::before,
    &::after {
        position: absolute;
        display: block;
        width: 100%;
        height: $scroll-shadow-height * 2;
        border-radius: inherit;
        opacity: 0;
        transition: opacity .1s ease-in-out;
        content: '';
    }

    // default
    &::before {
        top: -$scroll-shadow-height * 2;
        z-index: 20;
        box-shadow: 0 $scroll-shadow-height $scroll-shadow-height -2px rgba(0, 0, 0, .2);
    }

    &::after {
        bottom: -$scroll-shadow-height * 2;
        z-index: 30;
        box-shadow: 0 (-$scroll-shadow-height) $scroll-shadow-height -2px rgba(0, 0, 0, .2);
    }

    &.style--cover {
        &::before,
        &::after {
            width: 110%;
            margin: 0 -5%;
        }
    }

    &.style--contain {
        &::before {
            top: -($scroll-shadow-height + 10px);
        }

        &::after {
            bottom: -($scroll-shadow-height + 10px);
        }
    }

    &.is-showing-top-shadow::before {
        opacity: 1;
    }

    &.is-showing-bottom-shadow::after {
        opacity: 1;
    }
}
