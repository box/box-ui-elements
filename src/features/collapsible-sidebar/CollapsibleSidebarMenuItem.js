// @flow strict
import * as React from 'react';
import classNames from 'classnames';
import styled from 'styled-components';
import Tooltip from '../../components/tooltip';
import { useIsContentOverflowed } from '../../utils/dom';
import CollapsibleSidebarContext from './CollapsibleSidebarContext';

const StyledMenuItem = styled.div`
    position: relative;

    /* hover styles for link so that hovering both action icon and link
    will have hover effect over whole container */
    &:hover a {
        background-color: ${({ theme }) => theme.primary.backgroundHover};
    }

    &:hover a.is-currentPage {
        background-color: ${({ theme }) => theme.primary.backgroundActive};
    }

    body.is-move-dragging & a:hover {
        /* if an item is being dragged to left nav, keep menu item defaults (don't highlight) */
        background-color: ${({ theme }) => theme.primary.background};

        .bdl-CollapsibleSidebar-menuItemIcon {
            opacity: 0.7;
        }
    }

    .bdl-CollapsibleSidebar-menuItemActionContainer {
        position: absolute;
        top: 12px;
        right: 8px;
        padding: 0;
        opacity: 0;

        &:hover,
        &:focus-within {
            opacity: 1;
        }
    }
`;

const StyledIconWrapper = styled.span`
    line-height: 0; /* let inner svg set the height */
    opacity: 0.7;

    & path,
    & .fill-color {
        fill: ${({ theme }) => theme.primary.foreground};
    }

    a:active &,
    a:hover &,
    a:focus &,
    a.is-currentPage & {
        opacity: 1;
    }
`;

const StyledMenuItemLabel = styled.span`
    flex-grow: 1;
    overflow: hidden;
    color: ${({ theme }) => theme.primary.foreground};
    text-overflow: ellipsis;
    opacity: 0.85;

    &.overflow-action {
        margin-right: 12px;
    }

    a:active &,
    a:hover &,
    a:focus &,
    a.is-currentPage & {
        opacity: 1;
    }
`;

// {...rest} props will go here, such as `as` prop to adjust the component name.
// In most cases the consumer will want the tag to use a `Link` instead of `a`.
const StyledLink = styled.a`
    display: flex;
    align-items: center;
    height: ${({ theme }) => theme.base.gridUnitPx * 10}px;
    padding: 0 ${({ theme }) => theme.base.gridUnitPx * 3}px;
    overflow-x: hidden;
    color: ${({ theme }) => theme.primary.foreground};
    font-weight: bold;
    white-space: nowrap;
    border: 1px solid transparent;
    border-radius: ${({ theme }) => theme.base.gridUnitPx * 2}px;
    transition: background-color 0.15s cubic-bezier(0.215, 0.61, 0.355, 1);

    &:hover,
    &:active,
    &:focus,
    &.is-currentPage {
        .bdl-CollapsibleSidebar-menuItemIcon,
        .bdl-CollapsibleSidebar-menuItemLabel {
            opacity: 1;
        }
    }

    &:focus {
        border-color: ${({ theme }) => theme.primary.foreground};
        outline: none;
    }

    &:focus:active {
        border-color: transparent;
    }

    &.is-currentPage {
        background-color: ${({ theme }) => theme.primary.backgroundActive};
    }

    &.is-currentPage:active {
        border-color: ${({ theme }) => theme.primary.foreground};
    }

    .bdl-CollapsibleSidebar-menuItemIcon + .bdl-CollapsibleSidebar-menuItemLabel {
        margin-left: 16px;
    }

    &.show-overflowAction + .bdl-CollapsibleSidebar-menuItemActionContainer,
    &:focus + .bdl-CollapsibleSidebar-menuItemActionContainer,
    &:hover + .bdl-CollapsibleSidebar-menuItemActionContainer {
        opacity: 1;
    }
`;

type Props = {
    /** Additional classes */
    className?: string,
    /** Content to be  displayed next to item label. */
    content?: React.Node,
    icon?: React.Node,
    linkClassName?: string,
    overflowAction?: React.Node,
    shouldHideTooltip?: boolean,
    /** Default is to always show */
    showOverflowAction?: 'hover' | 'always',
    text?: string,
};

function CollapsibleSidebarMenuItem(props: Props) {
    const {
        className,
        content,
        icon,
        linkClassName,
        overflowAction,
        showOverflowAction,
        text,
        shouldHideTooltip = false,
        ...rest
    } = props;
    const textRef: { current: null | HTMLElement } = React.useRef(null);
    const isTextOverflowed = useIsContentOverflowed(textRef);
    const { isScrolling } = React.useContext(CollapsibleSidebarContext);
    const isShowOverflowActionOnHover = showOverflowAction === 'hover';
    const menuItemClassName = classNames('bdl-CollapsibleSidebar-menuItem', className);
    const menuItemLinkClassName = classNames('bdl-CollapsibleSidebar-menuItemLink', linkClassName, {
        'show-overflowAction': !isShowOverflowActionOnHover,
    });
    const menuItemLabelClassName = classNames('bdl-CollapsibleSidebar-menuItemLabel', {
        'overflow-action': overflowAction,
    });

    const renderMenuItem = () => {
        return (
            <StyledMenuItem className={menuItemClassName}>
                <StyledLink className={menuItemLinkClassName} {...rest}>
                    {icon && (
                        <StyledIconWrapper className="bdl-CollapsibleSidebar-menuItemIcon">{icon}</StyledIconWrapper>
                    )}
                    {text && (
                        <StyledMenuItemLabel ref={textRef} className={menuItemLabelClassName}>
                            {text}
                        </StyledMenuItemLabel>
                    )}
                    {content}
                </StyledLink>
                {overflowAction && (
                    <span className="bdl-CollapsibleSidebar-menuItemActionContainer">{overflowAction}</span>
                )}
            </StyledMenuItem>
        );
    };

    if (isScrolling) {
        return renderMenuItem();
    }

    return (
        <Tooltip
            className={classNames('bdl-CollapsibleSidebar-menuItemToolTip')}
            isDisabled={!isTextOverflowed}
            isShown={isScrolling || shouldHideTooltip ? false : undefined}
            isTabbable={false}
            position="middle-right"
            text={text}
        >
            {renderMenuItem()}
        </Tooltip>
    );
}

export default CollapsibleSidebarMenuItem;
