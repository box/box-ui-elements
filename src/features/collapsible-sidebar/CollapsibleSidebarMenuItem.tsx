/**
 * A sidebar component that supports collapsed/expanded state for responsive sizing.
 * This component should be moved into BUIE when complete.
 * This component should NOT contain any reference to EUA specific patterns like Immutables and redux containers.
 */

import * as React from 'react';
import classNames from 'classnames';
import styled from 'styled-components';
import Tooltip, { TooltipPosition } from '../../components/tooltip';
import { useIsContentOverflowed } from '../../utils/dom';
import CollapsibleSidebarContext from './CollapsibleSidebarContext';

import './CollapsibleSidebarMenuItem.scss';

const StyledMenuItem = styled.div`
    border: 1px solid transparent;
    color: ${props => props.theme.primary.foreground};

    a:hover:not(.is-currentPage) & {
        background-color: ${props => props.theme.primary.backgroundHover};
    }

    body.is-move-dragging a:hover & {
        // if an item is being dragged to left nav, keep menu item defaults (don't highlight)
        background-color: ${props => props.theme.primary.background};
    }

    a:focus & {
        border-color: ${props => props.theme.primary.foreground};
    }

    a:focus:active & {
        border-color: transparent;
    }

    a.is-currentPage & {
        background-color: ${props => props.theme.primary.backgroundActive};
    }

    a.is-currentPage:active & {
        border-color: ${props => props.theme.primary.foreground};
    }
`;

const StyledIconWrapper = styled.span`
    & path,
    & .fill-color {
        fill: ${props => props.theme.primary.foreground};
    }

    a:active &,
    a:hover &,
    a:focus &,
    a.is-currentPage & {
        opacity: 1;
    }
`;

const StyledMenuItemLabel = styled.span`
    color: ${props => props.theme.primary.foreground};

    a:active &,
    a:hover &,
    a:focus &,
    a.is-currentPage & {
        opacity: 1;
    }
`;

type Props = {
    /** Additional classes */
    className?: string;
    icon?: React.ReactNode;
    overflowAction?: React.ReactNode;
    text?: string;
};
function CollapsibleSidebarMenuItem(props: Props) {
    const { className, icon, overflowAction, text, ...rest } = props;
    const textRef = React.useRef<HTMLElement>(null);
    const isTextOverflowed = useIsContentOverflowed(textRef);
    const { isScrolling } = React.useContext(CollapsibleSidebarContext);

    const renderMenuItem = () => {
        return (
            <StyledMenuItem className={classNames('bdl-CollapsibleSidebar-menuItem', className)} {...rest}>
                {icon && <StyledIconWrapper className="bdl-CollapsibleSidebar-menuItemIcon">{icon}</StyledIconWrapper>}
                {text && (
                    <StyledMenuItemLabel ref={textRef} className="bdl-CollapsibleSidebar-menuItemLabel">
                        {text}
                    </StyledMenuItemLabel>
                )}
                {overflowAction}
            </StyledMenuItem>
        );
    };

    if (isScrolling) {
        return renderMenuItem();
    }

    return (
        <Tooltip
            className={classNames('bdl-CollapsibleSidebar-menuItemToolTip')}
            isDisabled={!isTextOverflowed}
            isShown={isScrolling ? false : undefined}
            isTabbable={false}
            position={TooltipPosition.MIDDLE_RIGHT}
            text={text}
        >
            {renderMenuItem()}
        </Tooltip>
    );
}

export default CollapsibleSidebarMenuItem;
