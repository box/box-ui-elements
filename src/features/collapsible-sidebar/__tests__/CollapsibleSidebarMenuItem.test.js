import * as React from 'react';
import Tooltip from '../../../components/tooltip';
import { mountConnected } from '../../../test-utils/enzyme';

import * as libDom from '../../../utils/dom';

import CollapsibleSidebarMenuItem from '../CollapsibleSidebarMenuItem';
import CollapsibleSidebarContext from '../CollapsibleSidebarContext';

jest.mock('../../../utils/dom', () => ({ useIsContentOverflowed: jest.fn() }));

describe('components/core/collapsible-sidebar/__tests__/CollapsibleSidebarMenuItem', () => {
    const getWrapper = (props, { isScrolling = false } = {}) => {
        return mountConnected(
            <CollapsibleSidebarContext.Provider value={{ isScrolling }}>
                <CollapsibleSidebarMenuItem {...props} />
            </CollapsibleSidebarContext.Provider>,
        );
    };

    beforeEach(() => {
        libDom.useIsContentOverflowed.mockReturnValue(false);
    });

    test('matches snapshot', () => {
        const wrapper = getWrapper(
            {
                className: 'foo',
                text: 'bar',
                icon: 'bold',
            },
            { isScrolling: false },
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should show custom content when content is passed', () => {
        const testContent = 'Custom Content';
        const wrapper = getWrapper(
            {
                className: 'foo',
                content: <div className="custom-div">{testContent}</div>,
                icon: 'bold',
                text: 'bar',
            },
            { isScrolling: false },
        );

        expect(wrapper.find('.custom-div')).toHaveLength(1);
        expect(wrapper.find('.custom-div').text()).toBe(testContent);
        expect(wrapper.find('span.bdl-CollapsibleSidebar-menuItemLabel')).toHaveLength(1);
    });

    test('should allow tooltip if text is overflowing and not scrolling', () => {
        libDom.useIsContentOverflowed.mockReturnValue(true);

        const wrapper = getWrapper({ text: 'bar', icon: 'bold' }, { isScrolling: false });

        expect(wrapper.find('.bdl-CollapsibleSidebar-menuItemToolTip').length).toBe(1);
        expect(wrapper.find(Tooltip).prop('isShown')).toBe(undefined); // can show on hover
        expect(wrapper.find(Tooltip).prop('isDisabled')).toBe(false);
    });

    test('should not render visible tooltip if text is overflowing and scrolling', () => {
        libDom.useIsContentOverflowed.mockReturnValue(true);

        const wrapper = getWrapper({ text: 'bar', icon: 'bold' }, { isScrolling: true });

        expect(wrapper.find('.bdl-CollapsibleSidebar-menuItemToolTip').length).toBe(0);
    });

    test('should spread props to inner element', () => {
        const mouseEvent = jest.fn();

        const wrapper = getWrapper({
            className: 'foo',
            text: 'bar',
            icon: 'bold',
            onMouseOver: mouseEvent,
        });

        const classNameTarget = wrapper.find('CollapsibleSidebarMenuItem__StyledMenuItem');
        const restTarget = wrapper.find('a');
        expect(restTarget.prop('onMouseOver')).toBe(mouseEvent);
        expect(classNameTarget.prop('className')).toBe('bdl-CollapsibleSidebar-menuItem foo');
    });

    test('should not show overflow action container', () => {
        const wrapper = getWrapper({
            className: 'foo',
            text: 'bar',
            icon: 'bold',
        });
        expect(wrapper.find('.bdl-CollapsibleSidebar-menuItemActionContainer').length).toBe(0);
    });

    test('should show overflow action by default', () => {
        const wrapper = getWrapper({
            className: 'foo',
            text: 'bar',
            icon: 'bold',
            overflowAction: <div>Hi</div>,
        });
        expect(wrapper.find('.show-overflowAction').length).toBe(2);
        expect(wrapper.find('.bdl-CollapsibleSidebar-menuItemActionContainer').length).toBe(1);
    });

    test('should show overflow action on hover if showAction is set to hover', () => {
        const wrapper = getWrapper({
            className: 'foo',
            text: 'bar',
            icon: 'bold',
            overflowAction: <div>Hi</div>,
            showOverflowAction: 'hover',
        });
        expect(wrapper.find('.show-overflowAction').length).toBe(0);
        expect(wrapper.find('.bdl-CollapsibleSidebar-menuItemActionContainer').length).toBe(1);
    });

    test('should show link class name when it is set', () => {
        const wrapper = getWrapper({
            className: 'foo',
            text: 'bar',
            icon: 'bold',
            linkClassName: 'is-currentPage',
        });
        expect(wrapper.find('CollapsibleSidebarMenuItem__StyledLink.is-currentPage').length).toBe(1);
    });
});
