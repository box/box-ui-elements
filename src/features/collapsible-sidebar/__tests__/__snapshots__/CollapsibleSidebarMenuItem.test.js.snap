// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/core/collapsible-sidebar/__tests__/CollapsibleSidebarMenuItem matches snapshot 1`] = `
<CollapsibleSidebarMenuItem
  className="foo"
  icon="bold"
  text="bar"
>
  <Tooltip
    className="bdl-CollapsibleSidebar-menuItemToolTip"
    constrainToScrollParent={false}
    constrainToWindow={true}
    isDisabled={true}
    isTabbable={false}
    position="middle-right"
    text="bar"
    theme="default"
  >
    <CollapsibleSidebarMenuItem__StyledMenuItem
      className="bdl-CollapsibleSidebar-menuItem foo"
    >
      <div
        className="CollapsibleSidebarMenuItem__StyledMenuItem-sc-1fmdq8m-0 cQLRdi bdl-CollapsibleSidebar-menuItem foo"
      >
        <CollapsibleSidebarMenuItem__StyledLink
          className="bdl-CollapsibleSidebar-menuItemLink show-overflowAction"
        >
          <a
            className="CollapsibleSidebarMenuItem__StyledLink-sc-1fmdq8m-3 OZzby bdl-CollapsibleSidebar-menuItemLink show-overflowAction"
          >
            <CollapsibleSidebarMenuItem__StyledIconWrapper
              className="bdl-CollapsibleSidebar-menuItemIcon"
            >
              <span
                className="CollapsibleSidebarMenuItem__StyledIconWrapper-sc-1fmdq8m-1 ivMWNp bdl-CollapsibleSidebar-menuItemIcon"
              >
                bold
              </span>
            </CollapsibleSidebarMenuItem__StyledIconWrapper>
            <CollapsibleSidebarMenuItem__StyledMenuItemLabel
              className="bdl-CollapsibleSidebar-menuItemLabel"
            >
              <span
                className="CollapsibleSidebarMenuItem__StyledMenuItemLabel-sc-1fmdq8m-2 khOLhH bdl-CollapsibleSidebar-menuItemLabel"
              >
                bar
              </span>
            </CollapsibleSidebarMenuItem__StyledMenuItemLabel>
          </a>
        </CollapsibleSidebarMenuItem__StyledLink>
      </div>
    </CollapsibleSidebarMenuItem__StyledMenuItem>
  </Tooltip>
</CollapsibleSidebarMenuItem>
`;
